src/methods/scgpt_fine_tuned/script.py

```diff
@@ -0,0 +1,503 @@
+  input_emb_style="continuous",
+  n_input_bins=hyperparameters["n_bins"],
+  cell_emb_style="cls",
+  mvc_decoder_style="inner product",
+  ecs_threshold=hyperparameters["ecs_threshold"],
+  explicit_zero_prob=explicit_zero_prob,
+  use_fast_transformer=hyperparameters["fast_transformer"],
+  fast_transformer_backend="flash", # "linear" or "flash"
+  pre_norm=hyperparameters["pre_norm"],
+)
+
+if model_file is not None:
+  # only load params that are in the model and match the size
+  map_location = torch.device('cpu') if device == 'cpu' else None
+  model_dict = model.state_dict()
+  pretrained_dict = torch.load((model_file), map_location=map_location)
+  pretrained_dict = {
+    k: v
+    for k, v in pretrained_dict.items()
+    if k in model_dict and v.shape == model_dict[k].shape
+  }
+  for k, v in pretrained_dict.items():
+    logger.info(f"Loading params {k} with shape {v.shape}")
+  model_dict.update(pretrained_dict)
+  model.load_state_dict(model_dict)
+
+pre_freeze_param_count = sum(dict((p.data_ptr(), p.numel()) for p in model.parameters() if p.requires_grad).values())
+
+# Freeze all pre-decoder weights
+for name, para in model.named_parameters():  
+  if hyperparameters["freeze"] and "encoder" in name and "transformer_encoder" not in name:
+    para.requires_grad = False
+
+post_freeze_param_count = sum(dict((p.data_ptr(), p.numel()) for p in model.parameters() if p.requires_grad).values())
+
+logger.info(f"Total Pre freeze Params {(pre_freeze_param_count )}")
+logger.info(f"Total Post freeze Params {(post_freeze_param_count )}")
+
+model.to(device)
+
+if training_settings["ADV"]:
+  discriminator = scgpt.model.AdversarialDiscriminator(
+    d_model=embsize,
+    n_cls=num_batch_types,
+  ).to(device)
+else: 
+  discriminator = None
+
+criterion = scgpt.loss.masked_mse_loss
+criterion_cls = torch.nn.CrossEntropyLoss()
+criterion_dab = torch.nn.CrossEntropyLoss()
+optimizer = torch.optim.Adam(model.parameters(), lr=hyperparameters["lr"], eps=1e-4 if hyperparameters["amp"] else 1e-8)
+scheduler = torch.optim.lr_scheduler.StepLR(optimizer, schedule_interval, gamma=hyperparameters["schedule_ratio"])
+
+if DAB_separate_optim:
+  optimizer_dab = torch.optim.Adam(model.parameters(), lr=hyperparameters["lr"])
+  scheduler_dab = torch.optim.lr_scheduler.StepLR(
+    optimizer_dab, schedule_interval, gamma=hyperparameters["schedule_ratio"]
+  )
+if training_settings["ADV"]:
+  criterion_adv = torch.nn.CrossEntropyLoss()  # consider using label smoothing
+  optimizer_E = torch.optim.Adam(model.parameters(), lr=lr_ADV)
+  scheduler_E = torch.optim.lr_scheduler.StepLR(
+    optimizer_E, schedule_interval, gamma=hyperparameters["schedule_ratio"]
+  )
+  optimizer_D = torch.optim.Adam(discriminator.parameters(), lr=lr_ADV)
+  scheduler_D = torch.optim.lr_scheduler.StepLR(
+    optimizer_D, schedule_interval, gamma=hyperparameters["schedule_ratio"]
+  )
+else:
+  criterion_adv = None
+  optimizer_E = None
+  optimizer_D = None
```

Comment:
Could some of the optional parts like this be simplified/removed if we only every use one branch?